cpp/src/replace/replace.cu

```diff
@@ -20,123 +20,265 @@
+   * @param[in] d_replacement_values Device array with the new values
+   * @param[in] replacement_valid Valid mask associated with d_replacement_values
    * 
    * @returns
    */
   /* ----------------------------------------------------------------------------*/
-  template <class T>
+  template <class T,
+            bool input_has_nulls, bool replacement_has_nulls>
   __global__
-  void replace_kernel(T*                          d_col_data,
-                      gdf_size_type                      nrows,
-                      thrust::device_ptr<const T> old_values_begin,
-                      thrust::device_ptr<const T> old_values_end,
-                      const T*                    d_new_values)
+  void replace_kernel(const T* __restrict__           input_data,
+                      bit_mask_t const * __restrict__ input_valid,
+                      T * __restrict__          output_data,
+                      bit_mask_t * __restrict__ output_valid,
+                      gdf_size_type *           output_null_count,
+                      gdf_size_type             nrows,
+                      thrust::device_ptr<const T> values_to_replace_begin,
+                      thrust::device_ptr<const T> values_to_replace_end,
+                      const T* __restrict__           d_replacement_values,
+                      bit_mask_t const * __restrict__ replacement_valid)
   {
-    gdf_size_type i = blockIdx.x * blockDim.x + threadIdx.x;
-    while(i < nrows)
-    {
-      auto found_ptr = thrust::find(thrust::seq, old_values_begin, old_values_end, d_col_data[i]);
+  gdf_size_type i = blockIdx.x * blockDim.x + threadIdx.x;
+
+  uint32_t active_mask = 0xffffffff;
+  active_mask = __ballot_sync(active_mask, i < nrows);
+
+  while (i < nrows) {
+    if (input_has_nulls) {
+      bool const input_is_valid{bit_mask::is_valid(input_valid, i)};
 
-      if (found_ptr != old_values_end) {
-          auto d = thrust::distance(old_values_begin, found_ptr);
-          d_col_data[i] = d_new_values[d];
+      uint32_t bitmask = __ballot_sync(active_mask, input_is_valid);
+
+      if (input_is_valid) {
+        auto found_ptr = thrust::find(thrust::seq, values_to_replace_begin,
+                                      values_to_replace_end, input_data[i]);
+        T new_value{0};
+        if (found_ptr != values_to_replace_end) {
+          auto d = thrust::distance(values_to_replace_begin, found_ptr);
+          new_value = d_replacement_values[d];
+          if (replacement_has_nulls) {
+            bitmask &= __ballot_sync(active_mask,
+                                 bit_mask::is_valid(replacement_valid, d));
+          }
+        } else {
+          new_value = input_data[i];
+        }
+        output_data[i] = new_value;
       }
 
-      i += blockDim.x * gridDim.x;
+      output_valid[bit_mask::bit_container_index(i)] = bitmask;
+      if (threadIdx.x % 32 == 0) {
+        atomicAdd(output_null_count, __popc(bitmask));
```

Comment:
I believe you want to sum the inversion of the bitmask to count nulls. Also, it will be faster to compute the null count per block, and then just have thread 0 of the block do the atomicAdd.

Reply:
See https://github.com/rapidsai/cudf/blob/a7068797b44353c7e2e65590a540b4f8219fa7ad/cpp/src/stream_compaction/copy_if.cuh#L176-L188